src/healthlake-mcp-server/awslabs/healthlake_mcp_server/fhir_operations.py

```diff
@@ -0,0 +1,695 @@
+from botocore.awsrequest import AWSRequest
+from botocore.exceptions import ClientError, NoCredentialsError
+from typing import Any, Dict, List, Optional, Tuple
+from urllib.parse import urljoin
+
+
+logger = logging.getLogger(__name__)
+
+# HealthLake API limits
+MAX_SEARCH_COUNT = 100  # Maximum number of resources per search request
+DATASTORE_ID_LENGTH = 32  # AWS HealthLake datastore ID length
+
+
+def validate_datastore_id(datastore_id: str) -> str:
+    """Validate AWS HealthLake datastore ID format."""
+    if not datastore_id or len(datastore_id) != DATASTORE_ID_LENGTH:
+        raise ValueError(f'Datastore ID must be {DATASTORE_ID_LENGTH} characters')
+    return datastore_id
+
+
+class FHIRSearchError(Exception):
+    """Exception raised for FHIR search parameter errors."""
+
+    def __init__(self, message: str, invalid_params: Optional[List[str]] = None):
+        """Initialize FHIR search error with message and optional invalid parameters."""
+        self.invalid_params = invalid_params or []
+        super().__init__(message)
+
+
+class AWSAuth(httpx.Auth):
+    """Custom AWS SigV4 authentication for httpx."""
+
+    def __init__(self, credentials, region: str, service: str = 'healthlake'):
+        """Initialize AWS SigV4 authentication with credentials and region."""
+        self.credentials = credentials
+        self.region = region
+        self.service = service
+
+    def auth_flow(self, request):
+        """Apply AWS SigV4 authentication to the request."""
+        # Preserve the original Content-Length if it exists
+        original_content_length = request.headers.get('content-length')
+
+        # Use minimal headers for signing - include Content-Length if present
+        headers = {
+            'Accept': 'application/fhir+json',
+            'Content-Type': 'application/fhir+json',
+            'Host': request.url.host,
+        }
+
+        # Add Content-Length to headers for signing if present
+        if original_content_length:
+            headers['Content-Length'] = original_content_length
+
+        # For GET requests, no body
+        body = None if request.method.upper() == 'GET' else request.content
+
+        # Create AWS request for signing
+        aws_request = AWSRequest(
+            method=request.method, url=str(request.url), data=body, headers=headers
+        )
+
+        # Sign the request
+        signer = SigV4Auth(self.credentials, self.service, self.region)
+        signer.add_auth(aws_request)
+
+        # Clear existing headers and set only the signed ones
+        request.headers.clear()
+        for key, value in aws_request.headers.items():
+            request.headers[key] = value
+
+        yield request
+
+
+class HealthLakeClient:
+    """Client for AWS HealthLake FHIR operations."""
+
+    def __init__(self, region_name: Optional[str] = None):
+        """Initialize the HealthLake client."""
+        try:
+            self.session = boto3.Session()
```

Comment:
Add the user agent extra for boto sessions:

`            user_agent_extra=f'awslabs/mcp/healthlake-mcp-server/__version__'`